fdbserver/DataDistributionQueue.actor.cpp

```diff
@@ -1807,20 +1958,171 @@ ACTOR Future<bool> rebalanceReadLoad(DDQueueData* self,
 		return false;
 	}
 
+	int priority = mapDataMovementToPriority(reason);
+
 	auto& metrics = metricsList[chosenIdx];
 	traceEvent->detail("ShardReadBandwidth", metrics.bytesReadPerKSecond);
 	//  Verify the shard is still in ShardsAffectedByTeamFailure
 	shards = self->shardsAffectedByTeamFailure->getShardsFor(
 	    ShardsAffectedByTeamFailure::Team(sourceTeam->getServerIDs(), primary));
-	for (int i = 0; i < shards.size(); i++) {
-		if (metrics.keys == shards[i]) {
-			self->output.send(RelocateShard(metrics.keys.get(), priority, RelocateReason::REBALANCE_READ));
-			self->updateLastAsSource(sourceTeam->getServerIDs());
-			return true;
+
+	RelocateOperation relocateOperation = selectRelocateOperation(
+	    self, traceEvent, reason, topKReply.minReadLoad, topKReply.maxReadLoad, srcLoad, metrics, bottomKReply);
+
+	traceEvent->detail("ZZXREBALANCE", 0)
+	    .detail("IsMountainChopper", isDataMovementForMountainChopper(reason))
+	    .detail("ShardSize", shards.size())
+	    .detail("Server", sourceTeam->getServerIDs())
+	    .detail("RelocateOperation", relocateOperation)
+	    .detail("ServerSize", sourceTeam->size())
+	    .detail("TeamSize", self->teamSize)
+	    .detail("SingleRegionTeamSize", self->singleRegionTeamSize);
+
+	// TODO@ZZX: delete
+	int nonZeroCount = 0;
+	for (int i = 0; i < metricsList.size(); i++) {
+		if (metricsList[i].bytesReadPerKSecond != 0) {
+			nonZeroCount++;
 		}
 	}
-	traceEvent->detail("SkipReason", "ShardNotPresent");
-	return false;
+	traceEvent->detail("MetricsSize", metricsList.size()).detail("NonZeroMetricSize", nonZeroCount);
+
+	if (relocateOperation == RelocateOperation::INVALID) {
+		// do nothing
+		traceEvent->detail("SkipReason", "NotMoveAuxTeam");
+		return false;
+	} else {
+		for (int i = 0; i < shards.size(); i++) {
+			if (metrics.keys == shards[i]) {
+				// TODO@ZZX: Actually, we can simply use RelocateShard().
+				if (relocateOperation == RelocateOperation::MULTIPLY_REPLICAS) {
```

Comment:
Can we use RelocateOperation and RelocateReason as parameters and reduce these to one block?

Reply:
Just use `self->output.send(RelocateShard(shard, moveReason, relocateReason, traceId));` in all case.